### ISSUE TITLE
Spec: Transpose

### ISSUE THREAD

**Issue by justinchuby** [onnx/onnx#4611]
# Ask a Question

### Question

It is unclear what types on tensors Transpose accepts from the spec. For example, does it accept 1d tensors as input? Is `perm` optional since it has words like "by default"? Does len(perm) need to be rank(input)?

### Further information
- Relevant Area: operators

- Is this issue related to a specific model?  

No

**Comment by jcwchen** [comment]
Hi @justinchuby,
I think these corner cases might just be undefined behavior.

> does it accept 1d or 0d tensors as inputs?

IIUC, at least from onnx shape inference perspective, they should be both fine case if not giving specified perm.

> So can perm be an empty list?

Shape inference will stop if the given perm is an empty list according to [this line](https://github.com/onnx/onnx/blob/6719d502c36496f9465b5cce071ce5942e60538a/onnx/defs/tensor/defs.cc#L1109).

Note that since current ONNX spec does not explicitly disallow these, probably the downstream runtime tools' behavior for these cases might also be undefined (either can work or cannot work).

> Is perm optional since it has words like "by default"? Does len(perm) need to be rank(input)?

`perm` is optional. Definitely for this case I agree we can add as `perm (optional): list of ints `, although the following description has explicitly mentioned "by default".

> Does len(perm) need to be rank(input)?

Currently shape inference does not forbid it, but I think we can add a check for it.

Feel free to submit a PR to resolve these ambiguity if you have bandwidth. Thanks!

**Comment by justinchuby** [comment]
If the input is 0d, I would expect perm to be an empty list. Would that be reasonable?

**Comment by jcwchen** [comment]
> If the input is 0d, I would expect perm to be an empty list. Would that be reasonable?

Makes sense to me if we want to explicitly add the restriction: `len(perm) needs to be rank(input)`.

**Comment by gramalingam** [comment]
> If the input is 0d, I would expect perm to be an empty list. Would that be reasonable?

Yes. I would say that the current behavior is a bug/limitation of the shape-inference code.

**Comment by gramalingam** [comment]
I also tried out the ONNX reference implementation. It fails for a scalar (0D) input ... I assume this is a fixable limitation.

**Comment by Pranav-s79** [comment]
Hi, I was interested on working on this issue.

My understanding is that the remaining work involves clarifying and testing `Transpose` behavior for scalar inputs, particularly whether a rank-zero tensor should accept `perm=[]`. I would start by reproducing the current shape-inference and reference-evaluator behavior, then add regression tests before making the smallest necessary implementation and specification changes.

Is this issue still available, and does that match the scope you would like addressed?
